compiler/optimizations/propagateDomainConstness.cpp

```diff
@@ -0,0 +1,249 @@
+    }
+  }
+}
+
+// help removing chpl__initCopy and chpl__autoCopy after resolution while
+// retaining constness information therein
+void removeInitOrAutoCopyPostResolution(CallExpr *call) {
+  Expr *parentExpr = call->parentExpr;
+  Expr *nextExpr = parentExpr->next;
+  bool addNoop = addNoopAsNextExprIfNeeded(parentExpr, nextExpr);
+
+  Symbol *argSym = NULL;
+  Type *argType = NULL;
+  if (SymExpr *se = toSymExpr(call->get(1))) {
+    argSym = se->symbol();
+    argType = argSym->getValType();
+  }
+  INT_ASSERT(argSym);
+  INT_ASSERT(argType);
+
+  call->replace(call->get(1)->remove());
+
+  // we removed the first argument already, so definedConst is the first
+  // argument now
+  SymExpr *secondArg = toSymExpr(call->get(1)->remove());
+  INT_ASSERT(secondArg);
+
+  Symbol *isConst = secondArg->symbol();
+  INT_ASSERT(isConst->type == dtBool);
+
+  if (argType->symbol->hasFlag(FLAG_DOMAIN)) {
+    Symbol *lhs = NULL;
+    if (CallExpr *parentCall = toCallExpr(parentExpr)) {
+      if (parentCall->isPrimitive(PRIM_MOVE)) {
+        if (SymExpr *lhsSE = toSymExpr(parentCall->get(1))) {
+          lhs = lhsSE->symbol();
+        }
+      }
+    }
+    INT_ASSERT(lhs);
+
+    if (!isShadowVarSymbol(lhs)) {
+      Expr *anchor = nextExpr;
+      setDefinedConstForDomainSymbol(lhs, nextExpr, anchor, isConst);
+
+      if (addNoop) {
+        nextExpr->remove();
+      }
+    }
+  }
+}
+
+static bool addNoopAsNextExprIfNeeded(Expr *baseExpr, Expr *&nextExpr) {
+  bool addNoop = false;
+
+  if (nextExpr == NULL) {
+    addNoop = true;
+  }
+  else if (CallExpr *nextCall = toCallExpr(nextExpr)) {
+    if (nextCall->isPrimitive(PRIM_RETURN)) {
+      addNoop = true;
+    }
+  }
+  else {
+    // I don't know what this next expression could be, so add a NOOP just in
+    // case
+    addNoop = true;
+  }
+
+  if (addNoop) {
+    CallExpr *noop = new CallExpr(PRIM_NOOP);
+    baseExpr->insertAfter(noop);
+    nextExpr = noop;
+  }
+
+  return addNoop;
+}
+                
+static void setDefinedConstForDefExprWithIfExprs(Expr* e) {
+  if (CallExpr *initCall = toCallExpr(e)) {
+    if (initCall->isNamed("chpl__buildDomainExpr")) {
+      initCall->argList.last()->replace(new SymExpr(gFalse));
+      return;
+    }
+    else if (initCall->isNamed("chpl__distributed")) {
+      initCall->get(3)->replace(new SymExpr(gFalse));
+      return;
+    }
+  }
+  else if (IfExpr *initIf = toIfExpr(e)) {
+    setDefinedConstForDefExprWithIfExprs(initIf->getThenStmt()->body.head);
+    setDefinedConstForDefExprWithIfExprs(initIf->getElseStmt()->body.head);
+  }
+}
+
+static VarSymbol *addFieldAccess(Symbol *receiver, const char *fieldName,
+                                 Expr *insBefore, Expr *&insAfter,
+                                 bool asRef) {
+  Type *receiverType = receiver->getValType();
+
+  AggregateType *aggType = toAggregateType(receiverType);
+  if (aggType == NULL) {
+    aggType = toDecoratedClassType(receiverType)->getCanonicalClass();
+  }
```

Comment:
I think it'd be better to write `AggregateType *aggType = toAggregateType(canonicalClassType(receiverType->getValType())); INT_ASSERT(aggType);`